src/zsys.c

```diff
@@ -347,6 +347,8 @@ zsys_shutdown (void)
 
     ZMUTEX_DESTROY (s_mutex);
 
+    s_initialized = false;
```

Comment:
Same as before, the mutex might be destroyed while still in use - this should be the first thing that happens, as before, to avoid any potential race

Reply:
Sorry, I don't get it :(

Reply:
s_initialized, being a boolean and thus below word size, is implicitly atomic. On top of that, most architectures have cache coherency guarantees for <= word size stores.
So even if you have multiple threads entering zsys_shutdown, if reading and changing s_initialized is the first thing that happens, you are pretty much guaranteed that only one will make it through, and the rest will exit immediately. At least on architectures people actually use.

By moving this store later on, this property is lost, and now it's possible that multiple threads can start processing zsys_shutdown at the same time, which might result in undefined behaviour due to the mutex, which gets destroyed while others might be blocked on it.

Reply:
@bluca thanks for explaining. What's your take on this solution?

Reply:
Looks good, thanks!